### DISCUSSION TITLE
What does the hash field mean in Nightingale callback data?

### DISCUSSION THREAD

**Discussion by LiangYang1018** [ccfos/nightingale#1172] [Q&A] [answered]
我自己测试发现：一条规则触发告警，告警恢复和再次触发，再次恢复的hash值都是一样的。
![image](https://user-images.githubusercontent.com/69097273/192479195-d39df729-6df7-4170-a8f4-417375b3d029.png)

那么是不是只要创建了一条告警规则，无论怎么修改这条规则的参数，这个hash值都是不变的？也就是说hash值唯一对应创建的一条规则，而不关联规则的具体内容？

**Comment by UlricQin** [comment, marked as answer]
hash字段的生成由两部分决定，一个是告警规则的ID，另一个是通过promql查询返回的vector的标签集。比如你配置了一条告警规则（规则ID假设是33）：

```
mem_used_percent > 80
```

如果有两个机器都触发了阈值，拿着这个 promql 去查询就会返回两条数据：

```
mem_used_percent{ident="10.2.3.4", region="bj"} 89
mem_used_percent{ident="10.2.3.5", region="bj"} 86
```

此时会产生两条告警事件，每个告警事件对应一个hash值，姑且可以这么理解：

```
hash1=md5(33+mem_used_percent{ident="10.2.3.4", region="bj"})
hash2=md5(33+mem_used_percent{ident="10.2.3.5", region="bj"})
```

如果未来10.2.3.4这个机器告警恢复了，hash值不会变，可以追踪到这个机器历史上所有的内存告警事件和内存恢复事件
